## String Formating

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

player = 'Thomas'

points = 33


'Last night, '+player+' scored '+str(points)+' points.'  # concatenation

f'Last night, {player} scored {points} points.'          # string formatting

There are three ways to perform string formatting.

The oldest method involves placeholders using the modulo % character.

An improved technique uses the .format() string method.

The newest method, introduced with Python 3.6, uses formatted string literals, called f-strings.

### Formatting with placeholders

You can use %s to inject strings into your print statements. The modulo % is referred to as a "string formatting operator".

In [1]:
print('I am going to inject %s here.'%'somthing')

I am going to inject somthing here.


In [3]:
# You can pass multiple items by placing them inside a tuple after the % operator.

print('I am going to inject %s text here, and %s text here.'%('some', 'more'))

I am going to inject some text here, and more text here.


In [4]:
# You can also pass variable names:
x, y = 'some', 'more'
print("I'm goint to inject %s text here, and  %s text here."%(x, y))

I'm goint to inject some text here, and  more text here.


### Format conversion methods.

It should be noted that two methods %s and %r convert any python object to a string using two separate methods: str() and repr(). %r and repr() deliver the string representation of the object, including quotation marks and any escape characters.

In [5]:
print("He said his name was %s." %'Fred')
print("He said his name was %r." %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


In [6]:
# As another example, \t inserts a tab into a string.

print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')


I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The %s operator converts whatever it sees into a string, including integers and floats. The %d operator converts numbers to integers first, without rounding. Note the difference below:

In [7]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)

I wrote 3.75 programs today.
I wrote 3 programs today.


### Padding and Precision of Floating point numbers

Floating point numbers use the format %5.2f. Here, 5 would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, .2f stands for how many numbers to show past the decimal point. Let's see some examples:

In [13]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [14]:
print("Floating point number: %1.0f" %(13.144))

Floating point number: 13


In [16]:
print('Floating point number: %1.5f' %(13.144))

Floating point number: 13.14400


In [17]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [18]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


### Multiple Formating

Nothing prohibits using more than one conversion tool in the same print statement:



In [19]:
print('First: %s, second: %5.2f, Third: %r' %('hi!', 3.1222, 'bye!'))

First: hi!, second:  3.12, Third: 'bye!'


### Formating with .format() method

A better way to format objects into your strings for print statements is with the string .format() method. The syntax is:

'String here {} then also {}'.format('something1','something2')

In [20]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


### The .format() method has several advantages over the %s placeholder method:

##### 1. Inserted objects can be called by index position:

In [21]:
print('The {2} {1} {0}'.format('fox', 'brown', 'quick'))

The quick brown fox


##### 2. Inserted object can be assaigned key word

In [23]:
print('First object: {a}, Second object: {b}, Third object: {c}'.format(a=1, b='Two', c=3.1))

First object: 1, Second object: Two, Third object: 3.1


##### 3. Inserted object can be reused avoiding duplucation

In [24]:
print('A %s saved is a %s earned.' %('penny', 'penny'))
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### Alignment, padding and precision with .format()

Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [28]:
print('{0:8} | {1:9}'.format('Fruit', 'Quality'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quality  
Apples   |       3.0
Oranges  |        10


By default, .format() aligns text to the left, numbers to the right. You can pass an optional <,^, or > to set a left, center or right alignment:

In [30]:
print('{0:<8} | {1:^8} | {2:>8}'.format('left', 'Center', 'Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11, 22, 33))

left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [31]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left', 'Center', 'Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11, 22, 33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [33]:
print("This is my ten-character, two-decimal number:%10.2f" %13.576)
print("This is my ten-character, two decimal number:{0:10.2f}".format(13.576))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

### Formated string literals(f-strings)

Introduced in Python 3.6, f-strings offer several benefits over the older .format() string method described above. For one, you can bring outside variables immediately into to the string rather than pass them as arguments through .format(var).

In [34]:
name = 'Fred'
print(f'He said his name is {name}.')

He said his name is Fred.


Pass !r to get the string representation:

In [35]:
print(f"He said his name is {name!r}")

He said his name is 'Fred'


### Floating Format follows "result: {value:{width}.{precision}}"

Where with the .format() method you might see {value:10.4f}, with f-strings this can become {value:{10}.{6}}

In [38]:

num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")


My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Note that with f-strings, precision refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [40]:
num = 23.45
print('My 10 character, four decimal number is:{0:10.4f}'.format(num))
print(f'My 10 character, four decimal number is:{num:{10}.{6}}')

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use .format() method syntax inside an f-string:

In [42]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"my 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
my 10 character, four decimal number is:   23.4500
